nvidia-docker module/package

averelld opened the pull request:
###### Motivation for this change
See #27999 and NixOS/patchelf/issues/44 for some reasons why this is so messy. The ldconfig patch should go in standard glibc, but it's probably better to do that separately later.

Quick test: `nvidia-docker run nvidia/cuda:10.0-runtime nvidia-smi`
###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Mic92 commented on pkgs/development/libraries/glibc/ldconfig-patchelf.patch:
```diff
@@ -0,0 +1,67 @@
++	  return 1;
++
++  dynamic_strings = (char *) (file_contents + dyn_entry->d_un.d_val - loadaddr);
++  check_ptr (dynamic_strings);
```
Is this something the glibc upstreamer could adopt?

averelld replied:
It's possible. I don't know if it's actually legal to put dynstr in a different segment, and I doubt a compiler would do it. Also, they have like a 10 page contribution checklist, but yes it should be tried.

Mic92 replied:
The contribution checklist looks doable as it is mostly how to format patches: https://sourceware.org/glibc/wiki/Contribution%20checklist

I can also submit a patch, but I would need a proper explanation of the change as required here: https://sourceware.org/glibc/wiki/Contribution%20checklist#Detailed_Explanation_of_the_Patch

averelld replied:
Alright, I've submitted a bug report: https://sourceware.org/bugzilla/show_bug.cgi?id=23964
and a patch to the mailing list, we'll see what happens.

Mic92 commented on pkgs/applications/virtualization/nvidia-docker/default.nix:
```diff
@@ -0,0 +1,74 @@
+  nvidia-container-runtime = fetchFromGitHub {
+    owner = "NVIDIA";
+    repo = "nvidia-container-runtime";
+    rev = "95836f333576730bf0cc5f3e6580a30495034926";
```
Was it possible to stick to released versions? https://github.com/NVIDIA/nvidia-container-runtime/releases

Mic92 commented on pkgs/applications/virtualization/nvidia-docker/default.nix:
```diff
@@ -0,0 +1,74 @@
+  src = fetchFromGitHub {
+    owner = "NVIDIA";
+    repo = "nvidia-docker";
+    rev = "5836467b537cd7c0bfd69a47327c0e62fa11ae29";
```
here too: https://github.com/NVIDIA/nvidia-docker/releases

Mic92 commented on pkgs/applications/virtualization/nvidia-docker/libnvc.nix:
```diff
@@ -0,0 +1,52 @@
+  src = fetchFromGitHub {
+    owner = "NVIDIA";
+    repo = "libnvidia-container";
+    rev = "deccb2801502675bd283c6936861814dbca99ecd";
```
and here: https://github.com/NVIDIA/libnvidia-container/releases

averelld replied:
Done

averelld commented:
@Mic92 Can we get this closer to merging somehow? I don't think any of the linked issues are likely to show activity soon, and the feature is gated behind an extra switch, so the risk of breaking stuff is low.

Mic92 commented on pkgs/applications/virtualization/nvidia-docker/default.nix:
```diff
@@ -0,0 +1,74 @@
+    homepage = https://github.com/NVIDIA/nvidia-docker;
+    description = "NVIDIA container runtime for Docker";
+    license = licenses.bsd3;
+    platforms = platforms.linux;
```
Can you restrict this to `x86_64-linux` I don't think it would work/make sense on aarch64.

Mic92 commented on pkgs/applications/virtualization/nvidia-docker/config.toml:
```diff
@@ -0,0 +1,13 @@
+#path = "/usr/bin/nvidia-container-cli"
+environment = []
+#debug = "/var/log/nvidia-container-runtime-hook.log"
+ldcache = "/tmp/ld.so.cache"
```
Is this directory secure? Since it is world-writeable an attacker might put a different file/symlink here. An alternative could be `/run/docker`

averelld replied:
I've changed it, in case it gets other uses in the future. Currently it had no real effect though, since the ldconfig resolving (of a mostly fixed list of nvidia-libs) is patched out and done instead via string construction of libname + loaded host-kernel nvidia module version.

averelld commented:
@Mic92 Comments addressed, and the branch has been rebased on top of master.

Infinisil commented on pkgs/top-level/all-packages.nix:
```diff
@@ -11817,6 +11817,13 @@ in
+  nvidia-docker = callPackage ../applications/virtualization/nvidia-docker {
+    inherit (linuxPackages) nvidia_x11;
+    glibc = (glibc.overrideAttrs (oldAttrs: {
+      patches = oldAttrs.patches ++ [ ../development/libraries/glibc/ldconfig-patchelf.patch ];
```
What's this patch for? Some comments would be appreciated. In any case you should move the apply of this patch inside the main nvidia-docker nix file, no need to do it here. Also no need to use `inherit (linuxPackages) nvidia_x11`, you can just take `linuxPackages` as an argument instead.

averelld replied:
This ldconfig patch is just a bug, and should probably be always applied, but that might break other things and would be a huge rebuild. But the linked glibc issue on sourceware might someday be looked at, so that might solve itself in the future.

Infinisil commented on pkgs/applications/virtualization/nvidia-docker/libnvc.nix:
```diff
@@ -0,0 +1,52 @@
+    sha256 = "082b1jvkg1dij6l7yvmi7vi1fdbmp7jw4snpkw76ggl4lcvn9g15";
+  };
+
+  patches = [ ./libnvc-ldconfig-and-path-fixes.patch ];
```
And what's this patch for? A comment would be nice.

averelld replied:
I've added some comments here. It's possible there might be a way to make ldconfig work directly, by specifying the search path, the cache file path etc. directly making it writable etc. But that is still two patches, and it has to find exactly these included nvidia_x11 libraries in this version anyway. And also the PATH would still need a patch or wrapping.

Infinisil commented on pkgs/applications/virtualization/nvidia-docker/libnvc.nix:
```diff
@@ -0,0 +1,52 @@
+  };
+
+  nvidia-modprobe = fetchurl {
+    url = "https://github.com/NVIDIA/nvidia-modprobe/archive/396.51.tar.gz";
```
No need to define this as an attribute to `mkDerivation`, define this in the `let in` above instead.

Infinisil commented on pkgs/applications/virtualization/nvidia-docker/default.nix:
```diff
@@ -0,0 +1,74 @@
+    sha256 = "1vx5m591mnvcb9vy0196x5lh3r8swjsk0fnlv5h62m7m4m07v6wx";
+  };
+
+  buildInputs = [ makeWrapper ];
```
Should be `nativeBuildInputs` (because `makeWrapper` is only used during the build)

Infinisil commented on pkgs/applications/virtualization/nvidia-docker/default.nix:
```diff
@@ -0,0 +1,80 @@
+    # .dynstr section is no longer in the first LOAD segment. See also
+    # https://sourceware.org/bugzilla/show_bug.cgi?id=23964 and
+    # https://github.com/NixOS/patchelf/issues/44
+    patches = oldAttrs.patches ++ [ ../../../development/libraries/glibc/ldconfig-patchelf.patch ];
```
Ah so the patch is from https://sourceware.org/bugzilla/attachment.cgi?id=11444&action=diff

You should then use `fetchpatch` instead of adding it as a file.

averelld replied:
This was the patch submitted because of this PR, but yes it's much better documented. Fixed.

Infinisil commented on pkgs/applications/virtualization/nvidia-docker/default.nix:
```diff
@@ -0,0 +1,80 @@
+    wrapProgram $out/bin/nvidia-container-cli \
+      --prefix LD_LIBRARY_PATH : /run/opengl-driver/lib:/run/opengl-driver-32/lib
+    cp ${./config.toml} $out/etc/config.toml
+    substituteInPlace $out/etc/config.toml --replace "@glibcbin@" ${glibc-ldconf.bin}
```
```nix
''
  substituteInPlace $out/etc/config.toml --subst-var-by glibcbin ${lib.getBin glibc-ldconf}
''
```

Infinisil commented on nixos/modules/virtualisation/docker.nix:
```diff
@@ -140,7 +149,8 @@ in
   config = mkIf cfg.enable (mkMerge [{
-      environment.systemPackages = [ cfg.package ];
+      environment.systemPackages = [ cfg.package ]
+        ++ (optional cfg.enableNvidia pkgs.nvidia-docker);
```
Nitpicks: no parens needed on this line

Infinisil commented on nixos/modules/virtualisation/docker.nix:
```diff
@@ -165,7 +176,8 @@ in
 
-        path = [ pkgs.kmod ] ++ (optional (cfg.storageDriver == "zfs") pkgs.zfs);
+        path = [ pkgs.kmod ] ++ (optional (cfg.storageDriver == "zfs") pkgs.zfs)
+          ++ (optional cfg.enableNvidia pkgs.nvidia-docker);
```
Not as many parens needed:
`.. ++ optional (cfg.storageDriver == "zfs") pkgs.zfs ++ optional cfg.enableNvidia ...`

Infinisil commented on pkgs/applications/virtualization/nvidia-docker/default.nix:
```diff
@@ -0,0 +1,80 @@
+    # https://sourceware.org/bugzilla/show_bug.cgi?id=23964 and
+    # https://github.com/NixOS/patchelf/issues/44
+    patches = oldAttrs.patches ++ [ ../../../development/libraries/glibc/ldconfig-patchelf.patch ];
+  }));
```
One level of  parenthesis can be removed.

Infinisil commented on pkgs/applications/virtualization/nvidia-docker/libnvc.nix:
```diff
@@ -0,0 +1,55 @@
+  buildInputs = [ libelf libcap libseccomp linuxPackages.nvidia_x11 ];
+
+  preInstall = ''
+    install -Dm755 -t $out/mnt ${linuxPackages.nvidia_x11.bin}/bin/* ${persistenced}/bin/*
```
Use `lib.getBin linuxPackages.nvidia_x11` instead.

Infinisil commented on pkgs/applications/virtualization/nvidia-docker/libnvc.nix:
```diff
@@ -0,0 +1,55 @@
+    touch deps/src/nvidia-modprobe-396.51/.download_stamp
+
+    # Replace searching PATH for nvidia-smi etc. with output/mnt copies
+    substituteInPlace src/nvc_info.c --replace "@out@" $out
```
Use `--subst-var-by out $out`

Infinisil commented on pkgs/applications/virtualization/nvidia-docker/libnvc.nix:
```diff
@@ -0,0 +1,55 @@
+  buildInputs = [ libelf libcap libseccomp linuxPackages.nvidia_x11 ];
+
+  preInstall = ''
+    install -Dm755 -t $out/mnt ${linuxPackages.nvidia_x11.bin}/bin/* ${persistenced}/bin/*
```
Why use $out/mnt for binaries?

averelld replied:
Those are binaries that are mounted into the container (and therefore patched to work inside), but that probably doesn't make sense for naming the source, so it's just /bin now.

averelld commented:
I realized that there are actually two problems left:
- This only works if hardware.opengl.driSupport32Bit must be true
- the linuxPackages.nvidia_x11 bin reference only works if the default kernel is in use.

How can I fix that? Especially accessing the actual kernelPackages from a package is probably not possible. So should i patch bins in a module and put them in /run or something?

averelld commented:
So I modified the original nvidia_x11 packages to provide usable binaries which are linked into /run similar to the opengl libraries. Doesn't look very elegant, but at least this way there's less chance of version incompatibilities.
Also I had to fix the patch, because /run/opengl-driver/lib can be a link or a directory depending on configuration.

averelld commented:
@Infinisil This could be re-reviewed, or I could rebase and squash the fixups on top of the current master.

Mic92 commented on nixos/modules/hardware/video/nvidia.nix:
```diff
@@ -172,6 +172,11 @@ in
       ++ lib.filter (p: p != null) [ nvidia_x11.persistenced ];
 
+    systemd.tmpfiles.rules = [
+      "L+ /run/nvidia-docker/bin - - - - ${nvidia_x11.bin}/origBin"
```
Is this useful for all users?

averelld replied:
No, you are right. I have gated it with the same flag.

Infinisil commented on pkgs/applications/virtualization/nvidia-docker/libnvc.nix:
```diff
@@ -0,0 +1,42 @@
+with lib; let
+
+  nvidia-modprobe = fetchurl {
+    url = "https://github.com/NVIDIA/nvidia-modprobe/archive/396.51.tar.gz";
```
You should parametrize this file by the nvidia version "396.51". It would be very easy to forget to update these numbers later down the file.

Infinisil commented on pkgs/applications/virtualization/nvidia-docker/libnvc.nix:
```diff
@@ -0,0 +1,42 @@
+with lib; let
+
+  nvidia-modprobe = fetchurl {
+    url = "https://github.com/NVIDIA/nvidia-modprobe/archive/396.51.tar.gz";
```
Since this is a github url, you can use `fetchFromGitHub` instead.

averelld replied:
Done, I copy it manually now and chmod below.

Infinisil commented on pkgs/applications/virtualization/nvidia-docker/libnvc.nix:
```diff
@@ -0,0 +1,42 @@
+  '';
+
+  buildInputs = [ libelf libcap libseccomp ];
+}
```
`meta` for this package is missing.

Infinisil commented on pkgs/applications/virtualization/nvidia-docker/libnvc.nix:
```diff
@@ -0,0 +1,42 @@
+  src = fetchFromGitHub {
+    owner = "NVIDIA";
+    repo = "libnvidia-container";
+    rev = "v1.0.0";
```
Use `rev = "v${version}"` instead.

Infinisil commented on pkgs/applications/virtualization/nvidia-docker/default.nix:
```diff
@@ -0,0 +1,84 @@
+  src = fetchFromGitHub {
+    owner = "NVIDIA";
+    repo = "nvidia-docker";
+    rev = "v2.0.3";
```
`rev = "v${version}";`

Infinisil commented on pkgs/applications/virtualization/nvidia-docker/default.nix:
```diff
@@ -0,0 +1,84 @@
+    })];
+  });
+
+  libnvidia-container = callPackage (import ./libnvc.nix) { };
```
Same as just `callPackage ./libnvc.nix { }`

Infinisil commented on nixos/modules/virtualisation/docker.nix:
```diff
@@ -194,7 +205,16 @@ in
     }
+    (mkIf cfg.enableNvidia {
+      environment.etc."nvidia-container-runtime/config.toml".text =
+        fileContents "${pkgs.nvidia-docker}/etc/config.toml";
```
I think that's just the same as
```nix
{
  environment.etc."nvidia-container-runtime/config.toml".source = "${pkgs.nvidia-docker}/etc/config.toml";
}
```

Infinisil commented on nixos/modules/virtualisation/docker.nix:
```diff
@@ -194,7 +205,16 @@ in
       };
+
+      assertions = [
+        { assertion = cfg.enableNvidia -> (config.hardware.opengl.driSupport32Bit or false);
```
Unnecessary parens

Infinisil commented on nixos/modules/hardware/video/nvidia.nix:
```diff
@@ -172,6 +172,11 @@ in
     environment.systemPackages = [ nvidia_x11.bin nvidia_x11.settings ]
       ++ lib.filter (p: p != null) [ nvidia_x11.persistenced ];
 
+    systemd.tmpfiles.rules = optional (config.virtualisation.docker.enableNvidia)
```
Unnecessary parens

averelld commented:
I also rebased, because the rebuild count went to 100+ for some reason, and maybe this makes it go back.

Infinisil commented on pkgs/applications/virtualization/nvidia-docker/libnvc.nix:
```diff
@@ -0,0 +1,53 @@
+  nvidia-modprobe = fetchFromGitHub {
+    owner = "NVIDIA";
+    repo = "nvidia-modprobe";
+    rev = "396.51";
```
Missed this one, `rev = modp-ver;`

averelld replied:
Doh.

Infinisil commented on pkgs/applications/virtualization/nvidia-docker/libnvc.nix:
```diff
@@ -0,0 +1,53 @@
+
+with lib; let
+
+  modp-ver = "396.51";
```
Actually, another thing regarding this: Doesn't this imply that this PR will only work for the 396.51 driver? Ideally we would somehow condition this on what the user uses in their NixOS system.

averelld replied:
No, thankfully not. This is just nixified from https://github.com/NVIDIA/libnvidia-container/blob/master/mk/nvidia-modprobe.mk#L9 basically. And for me at least that is not the version I'm running at (415.27) and 410 before. I'm assuming they only bump that when the API changes, not sure.
If I could use a fixed version, we wouldn't need anything in /run because it could just be referenced straight from the package. It would make everything much simpler, but then we need several versions of this package, and select those based on kernel/driver versions.

infinisil reviewed (commented):
The code looks good to me, but I haven't tested this myself (I do have an nvidia GPU). I'm not very familiar with docker though, can you give me some instructions to test this?

averelld commented:
Yes, basically this packages a simple wrapper around docker, `nvidia-docker` that adds `--runtime=nvidia` to the docker command, so the two are basically equivalent:
`nvidia-docker run nvidia/cuda:10.0-runtime nvidia-smi`
will pull and run that image, then run the mounted nvidia-smi bin.
Alternatively this line is taken directly from the tensorflow docker install page:
`docker run --runtime=nvidia -it --rm tensorflow/tensorflow:latest-gpu python -c "import tensorflow as tf; tf.enable_eager_execution(); print(tf.reduce_sum(tf.random_normal([1000, 1000])))"`
It should print something about your GPU when run.
You can also pull and run tensorflow/tensorflow:latest-gpu-jupyter and play around with that for experimentation.

averelld commented:
@Infinisil How's it looking? I think today was/is release branch-off?

averelld commented:
@FRidh You can ignore this, I accidentally pushed a commit in the wrong branch.

infinisil reviewed (approved):
Have tested with the command you gave, working well, thanks! If you want it in the stable release you can make a backport to 19.03.

Infinisil commented:
Oh actually before I merge, you should split the commit into 2: One for adding the package, and one for adding the NixOS option. And format your commit message according to the [Contribution Guidelines](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#submitting-changes), also linked to from the PR template.

Commit:
nvidia-docker: init at 2.0.3

nvidia_x11 and persistenced were modified to provide binaries which can be
mounted inside a docker-container to be executed there.

most ldconfig-based discovery of bundled nvidia libraries is patched out

ldconfig itself is patched to be able to deal with patchelf'ed libraries
See https://sourceware.org/bugzilla/show_bug.cgi?id=23964

Commit:
nixos/docker: add enableNvidia option

averelld commented:
Awesome. I'll make a backport PR once this is merged.

CMCDragonkai commented:
If I want run a container requiring GPU on NixOS with this MR, do I have still have to pass these parameters into the container?

```
  --env LD_LIBRARY_PATH='/lib/cuda' \
  --volume='/nix/store/gc779g2i2046prsxby8k1v92b0nfs2ix-nvidia-x11-390.67-4.17.6/lib:/lib/cuda:ro' \
```

I currently do this to ensure that the `libcuda.so` is available inside the container.

averelld commented:
Do you have an example image and command to test this?
I don't do that usually, but I also don't think libcuda is among the libs mounted into the container, including other distributions. If it turns out to be necessary we might adjust the runner accordingly.

CMCDragonkai commented:
Not with me, but when I created images using tensorflowWithCuda in 18.09, I would then run `docker run` with those parameters in order to make it work. If not, TF would complain of mismatched library versions. Basically the host `libcuda.so` had to be mounted into the container. However I thought using the new nvidia runtime would mean that it would automatically figure out to do this, since I didn't see any examples of those parameters being used by NVIDIA documentation.

averelld commented:
Ok, I checked the list of auto-mounted libs again: https://github.com/NVIDIA/libnvidia-container/blob/773b1954446b73921ce16919248c764ff62d29ad/src/nvc_info.c#L73
So yes, libcuda works without any special parameters. In general at least the "release" tensorflow images should work out of the box, for example like this:
`nvidia-docker run tensorflow/tensorflow:1.13.1-gpu-py3-jupyter python -c "import tensorflow as tf; tf.enable_eager_execution(); print(tf.reduce_sum(tf.random_normal([1000, 1000])))"` or see https://github.com/NixOS/nixpkgs/pull/51733#issuecomment-466389101 for alternate syntax.

CMCDragonkai commented:
@averelld I'll be trying that right now. Although I want to know that by `enableNvidia = true;` this means the default runtime for ALL docker containers (even ones that are not using GPUs) will use nvidia's runtime. Is this recommended? My reading of nvidia's runtime is that it actually is just a "modified" version of runc, whatever that means.

CMCDragonkai commented:
I just tried it now:

```
nvidia-docker run -it --rm \  ⚡(master) pts/13 18:01:18
  -p 0.0.0.0:55555:55555 \
  --expose 55555 \
  --volume='/tmp/image-classifier-zo8qv901/output:/settings:ro' \
  --env INVERTERS_API_HOST='0.0.0.0' \
  --env INVERTERS_API_PORT='55555' \
  --env INVERTERS_API_VERSION=0 \
  --env INVERTERS_API_WEIGHTS='/settings/weights.h5' \
  --env INVERTERS_API_CLASSES='/settings/classes.csv' \
  --env INVERTERS_API_STATS='/settings/images_redirected.json:/settings/thresholded.json:/settings/balanced.json:/settings/augmented.json' \
  --env INVERTERS_API_HISTORY='/settings/history.json' \
  --env INVERTERS_API_PARAMETERS='/settings/parameters.json' \
  --env NVIDIA_DRIVER_CAPABILITIES='compute,utility' \
  --env NVIDIA_VISIBLE_DEVICES=all \
  image-classifier-gpu:0gldxpli7c7x4rp1drknb56hsdiggwhx \
  /bin/image-classifier-server -v

Using TensorFlow backend.
2019-03-29 07:01:23.309784: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2019-03-29 07:01:23.315024: E tensorflow/stream_executor/cuda/cuda_driver.cc:397] failed call to cuInit: CUDA_ERROR_NO_DEVICE
2019-03-29 07:01:23.315063: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:158] retrieving CUDA diagnostic information for host: 303eb4ca0dd6
2019-03-29 07:01:23.315078: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:165] hostname: 303eb4ca0dd6
2019-03-29 07:01:23.315166: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:189] libcuda reported version is: 390.87.0
2019-03-29 07:01:23.315220: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:193] kernel reported version is: 415.27.0
2019-03-29 07:01:23.315237: E tensorflow/stream_executor/cuda/cuda_diagnostics.cc:303] kernel version 415.27.0 does not match DSO version 390.87.0 -- cannot find working devices in this configuration
Traceback (most recent call last):
  File "/nix/store/3x44q4zrqqhkb2bd4zvrdijzzgic9azq-image-classifier-0.0.1/bin/.image-classifier-server-wrapped", line 15, in <module>
    from image_classifier_api import app
  File "/nix/store/3x44q4zrqqhkb2bd4zvrdijzzgic9azq-image-classifier-0.0.1/lib/python3.6/site-packages/image_classifier_api/__init__.py", line 21, in <module>
    os.environ.get(env_prefix + 'GPU_FULL', 'false') == 'true')
  File "/nix/store/3x44q4zrqqhkb2bd4zvrdijzzgic9azq-image-classifier-0.0.1/lib/python3.6/site-packages/image_classifier/utils.py", line 33, in setup_gpu
    raise OSError('No GPUs available!')
OSError: No GPUs available!
```

It seems that there's still a problem, as that DSO error is exactly what I got before when I didn't mount the host's cuda driver into it.

CMCDragonkai commented:
Just thinking about this... this might be a nix specific issue, as the image that I'm trying to run used `dockerTools.buildImage` function which packages the `libcuda.so` into the container. Perhaps the nvidia runtime should be mounting the host's `libcuda.so` and writing the correct `LD_LIBRARY_PATH` to avoid this problem?

averelld commented:
Regarding the runtime, that is only selected when using the nvidia-docker wrapper or explicitly. The enable flag just prepares mountable libraries like libcuda and such.
You can check mounts in your container to see where libcuda is. Much of libnvidia-container relies on ldconfig which doesn't work with nix libraries, but adjusting LD_LIBRARY_PATH might be enough.
In the case of the error above, you can see it found an old libcuda, that probably wasn't mounted from the host, so there might just be an older version baked into the image somewhere.

CMCDragonkai commented:
Yes it is definitely the older version baked into the image. That's how `dockerTools.buildImage` works it packages up the nix closure into the container. However you seem you be saying that the nvidia runtime would somehow mount the right `libcuda.so`. Then perhaps if I know where it is mounted in the local container (or how to access it), I wouldn't need to separately mount the correct `libcuda.so` from the host.

averelld commented:
It does, yes, along with a selection of nvidia libs from `/run/opengl-driver{,-32}/lib/` and binaries like `nvidia-smi`. It could be here: https://github.com/NVIDIA/libnvidia-container/blob/e3a2035da5a44b8a83d9568b91a8a0b542ee15d5/src/common.h#L36 but you can just do `mount | grep libcuda` in the container to find the location. The runtime will also add symlinks, like so:
```
root@4e36fb6193d9:/# ll /usr/lib/x86_64-linux-gnu/libcuda*
lrwxrwxrwx 1 root root       12 Mar 29 10:40 /usr/lib/x86_64-linux-gnu/libcuda.so -> libcuda.so.1*
lrwxrwxrwx 1 root root       17 Mar 29 10:40 /usr/lib/x86_64-linux-gnu/libcuda.so.1 -> libcuda.so.418.43*
-rw-r--r-- 1 root root 15623896 Sep 12  2018 /usr/lib/x86_64-linux-gnu/libcuda.so.410.48
-r-xr-xr-x 1 root root 16033504 Jan  1  1970 /usr/lib/x86_64-linux-gnu/libcuda.so.418.43*

```
The ones in `/run/` are the libs you would mount anyway, the ones in the nix store could be the wrong version or compiled for the wrong kernel.

CMCDragonkai commented:
I found it. The NVIDIA docker on NixOS puts `libcuda.so` into `/usr/lib64`.

When I used bash to enter into my container, the only place that had libraries was `/usr/lib64` there was no other things like `/usr/lib/x86_64-linux-gnu`, there was not `/usr/lib` and there was no `/run` either. I think at minimum the NVIDIA docker container only produces the `/usr/lib64`. I wonder how your container has `/run/opengl-driver...` and `/usr/lib/x86_64-linux-gnu`. Note that my container image is built from `dockerTools.buildImage` which builds a container from scratch, not from an OS distro.

CMCDragonkai commented:
@averelld The above solution now works for my tensorflow container. However  I wanted to do a little test on a basic trivial container to see what the nvidia-docker integration does step by step. I ended up hitting an error that is unexpected.

```sh
# first I build a trivial container containing only bash and coreutils
docker load --input "$(
  nix-build \
    --no-out-link \
    -E 'with import <nixpkgs> {}; dockerTools.buildImage { name = "bash"; tag = "latest"; contents = [ bash coreutils ]; }'
)"

# then I run it with --runtime=nvidia (same result with nvidia-docker ...)
docker run --runtime=nvidia -it --rm \
  --env LD_LIBRARY_PATH='/usr/lib64' \
  --env NVIDIA_DRIVER_CAPABILITIES='compute' \
  --env NVIDIA_VISIBLE_DEVICES=all \
  bash:latest \
  /bin/bash
```

The result unfortunately is this error:

```
/nix/store/5lc85iln575fb8xwaxlxmp9z8jsd321q-docker-18.09.2/libexec/docker/docker: Error response from daemon: OCI runtime create failed: container_linux.go:344: starting container process caused "process_linux.go:424: container init caused \"process_linux.go:407: running prestart hook 1 caused \\\"error running hook: exit status 1, stdout: , stderr: exec command: [/nix/store/rzzc02akgdkq3fl94vmll0fqyg92195s-nvidia-docker-2.0.3/bin/nvidia-container-cli --load-kmods --ldcache=/tmp/ld.so.cache configure --ldconfig=@/nix/store/s9gx039ip2kycam82hzfy1i68kl906a1-glibc-2.27-bin/bin/ldconfig --device=all --compute --pid=15632 /var/lib/docker/devicemapper/mnt/d47e4b7194eac442d6ff4c232301466ba8bd432ff2e7f7d4580881c12371744f/rootfs]\\\\nnvidia-container-cli: ldcache error: process /nix/store/s9gx039ip2kycam82hzfy1i68kl906a1-glibc-2.27-bin/bin/ldconfig failed with error code: 1\\\\n\\\"\"": unknown.
```

averelld commented:
Regarding the paths above, those paths are image dependent, there is no "NVIDIA docker on NixOS" path unfortunately, and the `/run/opengl...` paths, are host paths *from* which the libraries are mounted.

The run command above apparently fails because there is no `/tmp`, which ldconfig writes to. If you add an appropriate mkdir to buildImage, like with runAsRoot it works.

CMCDragonkai commented:
Requires `/tmp` got it.

tbenst commented:
@CMCDragonkai do you have an example of using `buildImage` so that it works with nvidia-docker on NixOS? I just tried building

buildPytorch.nix:
```
let
  pkgs = import ./nixpkgs.nix;
  python-env = pkgs.python3.buildEnv.override {
    extraLibs = with pkgs.python3Packages; [
      pytorch
    ];
  };

in
pkgs.dockerTools.buildImage {
  name = "pytorch"; 
  contents = python-env; 
  runAsRoot = '' 
    #!${pkgs.runtimeShell}
    mkdir -p /data
  '';
  config = { 

    Cmd = [ "/bin/python" ];
    WorkingDir = "/data";
    Volumes = {
      "/data" = {};
    };
  };
}
```
but no GPU available
other files: https://gist.github.com/tbenst/a2ceffd8f5b2e5c646b4ca3516e7db60
